### DISCUSSION TITLE
Building MicroPython for Pico W

### DISCUSSION THREAD

**Discussion by shabaz123** [micropython/micropython#9627] [Hardware & Peripherals] [answered]
Hello,

I'm able to successfully build the MicroPython firmware for the original (non-W) Pico, using the zip/tar source files here:
[micropython.org/download](https://micropython.org/download/)

However, what's the procedure to build it for the Pico W?
For instance, if I wish to build the release version **1.19.1**, I have tried this:
`cmake .. -DMICROPY_BOARD-PICO_W`

It fails, because there is no **PICO_W** folder in **rp2/boards**

Since that didn't work, I tried this:

```
git clone https://github.com/micropython/micropython
cd micropython
git checkout tags/v1.19.1
make -C mpy-cross
cd ports/rp2
mkdir build
cd build
cmake .. -DMICROPY_BOARD-PICO_W
```

It fails, again because there's no **rp2/boards/PICO_W** in the v1.19.1 tagged code.

If I try to build the head, it fails because there's no pico-sdk and tinyusb etc., whereas if I build from the zip/tar file mentioned earlier, it all just works.

I don't know precisely what to do here. I have tried dumping the pico-sdk into the micropython/lib folder, but again there were errors (pasted below).

What's the easiest way to build v1.19.1 MicroPython for the Pico W? Anyone have a working set of commands I should type?

Many thanks!

```
shabaz@u18-nuc:~/development/micropython/micropython/ports/rp2/build$ cmake ..
PICO_SDK_PATH is /home/shabaz/development/micropython/micropython/lib/pico-sdk
Defaulting PICO_PLATFORM to rp2040 since not specified.
Defaulting PICO platform compiler to pico_arm_gcc since not specified.
PICO compiler is pico_arm_gcc
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
Build type is MinSizeRel
PICO target board is pico.
Using board configuration from /home/shabaz/development/micropython/micropython/lib/pico-sdk/src/boards/include/boards/pico.h
-- Found Python3: /usr/bin/python3.8 (found version "3.8.5") found components: Interpreter 
CMake Warning at /home/shabaz/development/micropython/micropython/lib/pico-sdk/src/rp2_common/tinyusb/CMakeLists.txt:14 (message):
  PICO_TINYUSB_PATH specified but content not present.


CMake Warning at /home/shabaz/development/micropython/micropython/lib/pico-sdk/src/rp2_common/pico_lwip/CMakeLists.txt:14 (message):
  PICO_LWIP_PATH specified but content not present.


Found User C Module(s): 
CMake Error at CMakeLists.txt:415 (pico_find_compiler):
  Unknown CMake command "pico_find_compiler".


-- Configuring incomplete, errors occurred!
See also "/home/shabaz/development/micropython/micropython/ports/rp2/build/CMakeFiles/CMakeOutput.log".
```

**Comment by robert-hh** [comment]
Instead of using the tarball it's better to clone the repository with `git clone https://github.com/micropython/micropython`. Then you get the full package with the most recent files.

**Reply by shabaz123** [reply]
Thanks! 
Unfortunately I'm missing the entire steps to build from the repo. These are the commands I'm using:

```
git clone https://github.com/micropython/micropython
cd micropython
git checkout tags/v1.19.1
make -C mpy-cross
cd ports/rp2
mkdir build
cd build
cmake .. -DMICROPY_BOARD-PICO_W
```

It fails due to no pico-sdk, so I did a git clone of pico-sdk into the **micropython/lib** folder, but that didn't resolve it (and it also reports ancillary projects missing like tinyusb. 

The original non-W Pico MicroPython build is easy to do from the tar file, but I cannot find instructions to build Pico and Pico W from the repo.

**Comment by jimmo** [comment]
@shabaz123  The Pico W did not exist at the time of the v1.19.1. That's why the board directory doesn't exist. You'll need to use a later revision (i.e. the current head of the master branch)

**Comment by robert-hh** [comment]
You may have missed the step_

`make submodules`

in the MicroPyhton directory. That will add all required submodules in the right version. No need to install them separately. And yes, better stay in the master branch and the most recent state.
See also the instruction and the README.md in the micropython directory and the director at the respective ports, if present. If you follow these instructions carefully, it works.

**Comment by sosi-deadeye** [comment, marked as answer]
Why so complicated? I use `make -j 16 BOARD=PICO_W` inside the port directory of **rp2** and **esp32**.
Sometimes `make submodules` is required, to get the new dependencies.

If the current working directory is _micropython_, then you can also use make from there:
```
make -C mpy-cross/ -j 16
make -C ports/rp2 BOARD=PICO_W clean
make -C ports/rp2 BOARD=PICO_W submodules
make -C ports/rp2 BOARD=PICO_W -j 16
```

Then ports/rp2/**build-PICO_W** is created. There you can find the **firmware.uf2**

**Reply by SmokeyBrandie** [reply]
The board has been renamed in the meantime to RPI_PICO_W. 

```
make -C mpy-cross/ -j 16
make -C ports/rp2 BOARD=RPI_PICO_W clean
make -C ports/rp2 BOARD=RPI_PICO_W submodules
make -C ports/rp2 BOARD=RPI_PICO_W -j 16
```

**Reply by scruss** [reply]
I'd refer to the [Building](https://github.com/micropython/micropython/tree/master/ports/rp2#building) instructions in the ports/rp2 folder. The method has changed a couple of times over the years, but this should be up to date.

**Comment by shabaz123** [comment]
Thank you, @robert-hh , @jimmo and @sosi-deadeye !
You were right. I was missing the submodules command.
I typed the commands as @sosi-deadeye lists, and the build worked.

Thanks!

```
MicroPython v1.19.1-544-g89b320737-dirty on 2022-10-15; Raspberry Pi Pico W with RP2040

Type "help()" for more information.

```
